irohad/ametsuchi/impl/flat_file/flat_file.cpp

```diff
@@ -16,132 +16,84 @@
+   *
+   * @param id - for conversion
+   * @return string repr of identifier
+   */
+  std::string id_to_name(Identifier id) {
+    std::ostringstream os;
+    os << std::setw(DIGIT_CAPACITY) << std::setfill('0') << id;
+    return os.str();
   }
-}
 
-/**
- * Checking consistency of storage for provided folder
- * If some block in the middle is missing all blocks following it are deleted
- * @param dump_dir - folder of storage
- * @return - last available identifier
- */
-nonstd::optional<Identifier> check_consistency(const std::string &dump_dir) {
-  auto log = logger::log("FLAT_FILE");
-
-  Identifier tmp_id = 0u;
-  if (dump_dir.empty()) {
-    log->error("check_consistency({}), not directory", dump_dir);
-    return nonstd::nullopt;
-  }
-  // Directory iterator:
-  struct dirent **namelist;
-  auto status = scandir(dump_dir.c_str(), &namelist, nullptr, alphasort);
-  if (status < 0) {
-    log->error("check_consistency({}), scandir error: {}", dump_dir, status);
-    return nonstd::nullopt;
-  }
-  if (status < 3) {
-    log->info("check_consistency({}), directory is empty", dump_dir);
-    auto n = static_cast<uint32_t>(status);
-    for (auto j = 0u; j < n; ++j) {
-      free(namelist[j]);
+  /**
+   * Checking consistency of storage for provided folder
+   * If some block in the middle is missing all blocks following it are deleted
+   * @param dump_dir - folder of storage
+   * @return - last available identifier
+   */
+  nonstd::optional<Identifier> check_consistency(const std::string &dump_dir) {
+    auto log = logger::log("FLAT_FILE");
+
+    if (dump_dir.empty()) {
+      log->error("check_consistency({}), not directory", dump_dir);
+      return nonstd::nullopt;
     }
-    free(namelist);
-    return 0;
-  }
-
-  auto n = static_cast<uint32_t>(status);
-  tmp_id = 2;
-
-  while (tmp_id < n and id_to_name(tmp_id - 1) == namelist[tmp_id]->d_name) {
-    ++tmp_id;
-  }
-  for (auto j = tmp_id; j < n; ++j) {
-    remove(dump_dir, namelist[j]->d_name);
-  }
 
-  for (auto j = 0u; j < n; ++j) {
-    free(namelist[j]);
+    auto const files = [&dump_dir] {
+      std::vector<boost::filesystem::path> ps;
+      std::copy(boost::filesystem::directory_iterator{dump_dir},
+                boost::filesystem::directory_iterator{},
+                std::back_inserter(ps));
+      std::sort(ps.begin(),
+                ps.end(),
+                [](const boost::filesystem::path &lhs,
+                   const boost::filesystem::path &rhs) {
+                  return lhs.compare(rhs) < 0;
+                });
+      return ps;
+    }();
+
+    auto const missing =
+        std::find_if(files.cbegin(),
+                     files.cend(),
+                     [id = 0](const boost::filesystem::path &p) mutable {
+                       ++id;
+                       return id_to_name(id) != p.filename();
+                     });
+    std::for_each(missing, files.cend(), [](const boost::filesystem::path &p) {
+      boost::filesystem::remove(p);
```

Comment:
Isn't this work as well?
`std::for_each(missing, files.cend(), boost::filesystem::remove);`

Reply:
No, the compiler can't deduce the type for the third argument because boost::filesystem::remove is overloaded.

Reply:
Quite strange, ok

Reply:
Yeah, sometimes overloading and template doesn't work well.
https://stackoverflow.com/questions/12638166/g-compiler-error-couldnt-deduce-template-parameter-funct